### ISSUE TITLE
Support for validation

### ISSUE THREAD

**Issue by tomasherman** [twilio/guardrail#231]
Hey,

are there any plans/interest in generating validation code for DTO objects? For example arrays that contain certain number of elements, strings with length constraints etc?

Thanks!

**Comment by blast-hardcheese** [comment]
Yep, just hasn't been a priority for us as of yet. I've been thinking of translating validation to https://github.com/fthomas/refined, though I've heard some folk are already using that through `x-scala-type` and some `imports`.

I suspect what will need to happen to support different libraries which can be emailed to provide different functionality is to decompose new and constrained algebras to represent _just_ validation, for instance, then enable or disable the `refined` interpreter. If others want different behavior, they can swap in new interpreters that use different libraries, or the noop one that just passes values through.

**Comment by tomasherman** [comment]
I'm thinking on about taking a stab at this...for our use case, we don't get to change the swagger definitions so we must handle it differently (we are unable to add `x-scala-type`)

I'm not really sure what you mean by this:

> I suspect what will need to happen to support different libraries which can be emailed to provide different functionality is to decompose new and constrained algebras to represent just validation, for instance, then enable or disable the refined interpreter. If others want different behavior, they can swap in new interpreters that use different libraries, or the noop one that just passes values through.

What i had in mind was to have two modes - `no validation` or `refined` validation which can be switched using cli arguments ... and for no validation, everything would work as is now but for `refined`, you would get types enriched with `refined` features

**Comment by blast-hardcheese** [comment]
whoops, `s/emailed/enabled/`:

> I suspect what will need to happen to support different libraries which can be enabled to provide different functionality is to...
> decompose new and constrained algebras to represent just validation, for instance...
> then enable or disable the refined interpreter.

This is effectively what you suggested 😄

> If others want different behavior, they can swap in new interpreters that use different libraries, or the noop one that just passes values through.

The noop one would just be the current behaviour.

The big point is coming up with a new `Free` algebra that represents the set of validation features, transforming the generated types on their way through the pipeline.

Fortunately, there's almost no reflection once generated (apart from some `case t"Iterable[$tpe]"` stuff during parameter transform), so once you construct the types, they won't be messed with.

**Comment by tomasherman** [comment]
Hey, it's me again ... so i spent a little time going through the code and i have some questions and insights.

1) Do we actually want an implementation that ignores validations, once we have support for them in place? It seems that only reason they are ignored is just the fact that guardrail doesn't support it. What point is then to design the code changes to allow them to be ignored? I think that if they are in swagger, there is no sense in giving people the power to skip generating them.
2) I think there are two possible approaches to validation
  - `refined` library, which is mighty cool and simplifies things to simply modifying type of param and everything else should work automatically, but won't work for java ... i guess we could bypass this limitation by enriching `ProtocolParameter` with field: `annotations: List[Annotation]` and then store annotations like [these][1] inside (@kelnos what do you think about it?)
  - or we can simply enrich `ProtocolParameter` with field `validations: List[L#FunctionLiteral]` which throws exception if value is invalid. This has the benefit of working for both java and scala but will require more boilerplate

Just as reminder, this is what ProtocolParameter looks like:

```scala
case class ProtocolParameter[L <: LA](term: L#MethodParameter,
                                      name: String,
                                      dep: Option[L#TermName],
                                      readOnlyKey: Option[String],
                                      emptyToNull: EmptyToNullBehaviour,
                                      defaultValue: Option[L#Term])
```

So, thoughts? If you still prefer new `Free` algebra, would you mind describing it a bit more? Perhaps few case classes or something? I can't really imagine it what it would look like, but i would like to work on this next as my boss is a bit sad that guardrail doesn't support this :)

Thanks!

[1]: https://www.baeldung.com/javax-validation

**Comment by kelnos** [comment]
Hey!  One problem I think we need to solve is around returning responses, and possibly enforcing spec file completeness.  Let's say I generally return some `ErrorResponse` class with error statuses.  Say it has three properties: (`http_status`, `error_code`, `error_message`).  Obviously when I am writing handler code in a server, and I want to respond with an error, I can construct an `ErrorResponse` instance and pass it back.

But with validations, we do the entire request/response cycle for the user, and their handler code never sees the request.  So one thing I'd like to consider is that, with validations enabled, the handler gets a method called `handleValidationFailure()` or something that allows the user to return a domain object (and custom status code?) on failures.

I'm about to hop on a plane; I have more thoughts but will write them up later.

**Comment by tomasherman** [comment]
That is good point! I haven't thought about that case ... my only experience is with java swagger implementation and what they do is i think simply returning 400 with some message. But i agree that having the ability to return custom object would be really nice. 

One thing we could do is instead of calling handlers with some domain request representation of type `A`, we would provide user with something like `ValidatedNel[String, A]` so that the user can decide what to do with it

**Comment by tomasherman** [comment]
@blast-hardcheese what do you think about the above? i have some free time finally to work on this but i need some direction ;)

**Comment by blast-hardcheese** [comment]
Reading scrollback to regain context...

**Comment by blast-hardcheese** [comment]
So, the problem with `ValidatedNel[String, A]` is the lack of context to the field that was the error; returning an error string prevents:
- Purpose tailored error messages
- Non-string error messages

Just throwing something up for discussion:

```scala
def doFoo(respond: FooResponse.type)(int: Int): F[FooResponse] = ???

case class ErrorIndex(path: List[ErrorPathElement], message: String)
def mapFooError(respond: FooResponse.type)(errorInfo: ErrorIndex): F[FooResponse.type] = handleError(respond.BadRequest(_))(errorInfo)
def handleBadRequest[A](body: Zzz => A)(errorIndex: ErrorIndex): F[A] = ???
```

The considerations:
- No violation of the spec, even in the error case
- Sane defaults, emitting a single error message if possible (? I'm ok for dropping this, just looking to ease onboarding)
- Similar structures for `BadRequest` defer to common implementations, so custom error builders can be all overridden at once

What do you think? I'd rather accomplish this via composition, but without requiring polymorphic source APIs similar to the magnet pattern for supplying different handlers, I don't have any good ideas.

An alternate approach would be to have explicit names per error body type, which would resolve the issue of needing to rely on function overloading for linking. That would be similar to what I've presented above, but with longer function names. An advantage would be lining up a clearer path for dynamic languages.

I'm very much open to future ideation on this, I don't think any particular implementation stands out currently.

**Comment by tomasherman** [comment]
you proposal looks reasonable to me ... i assume `handleError` and `handleBadRequest` is meant to be the same method? ... we probably want to be able to handle list of `ErrorIndex` so that we validate all fields at once

> An advantage would be lining up a clearer path for dynamic languages.

^--- are dynamic languages target for guardrail?


so the idea is to generate multiple `map???Error` and one `handleBadRequest`? perhaps it should be called `handleInvalidRequest`?

i'm not really sure how users would create one common handler in such scenario considering the type `???Response.BadRequest` will differ based on `???` - or am i missing something?
